### PR TITLE
Support building on windows

### PR DESCRIPTION
**Pull request by bhackett1024**
https://linear.app/replay/issue/RUN-1452

### CODE REVIEW

File: base/functional/callback_helpers.cc [hunk 1]
```diff
@@ -4,22 +4,29 @@
...
-static void* gRecordReplayValueFn;
+#else
+#include <windows.h>
+#endif
+
+static void* LookupRecordReplaySymbol(const char* name) {
+#ifndef _WIN32
+  void* fnptr = dlsym(RTLD_DEFAULT, name);
+#else
+  HMODULE module = GetModuleHandleA("windows-recordreplay.dll");
```
Comment (Domiii) [inline base/functional/callback_helpers.cc:17]: Can we use this as a push toward better code sharing between all our runtimes and runtime repos?
We are copy-and-pasting so much of this 😢 
And then, sometimes, there are some unnecessary changes between them, e.g. some names are different in `skia`, not sure if that is all that necessary?

File: base/lazy_instance_helpers.cc [hunk 2]
```diff
@@ -9,39 +9,45 @@
...
-
-namespace base {
-namespace internal {
-
-static void (*gRecordReplayBeginPassThroughEventsFn)();
+#else
+#include <windows.h>
+#endif
+
+static void* LookupRecordReplaySymbol(const char* name) {
```
Comment (Domiii) [inline base/lazy_instance_helpers.cc:18]: Here is another copy of `LookupRecordReplaySymbol`?
Code sharing 🙏

File: base/logging.cc [hunk 3]
```diff
@@ -122,21 +120,33 @@ typedef FILE* FileHandle;
...
 #endif
 
-static void (*gRecordReplayPrintFn)(const char*, va_list);
+#if !BUILDFLAG(IS_WIN)
+#include <dlfcn.h>
+#else
+#include <windows.h>
+#endif
+
+static void* LookupRecordReplaySymbol(const char* name) {
```
Comment (Domiii) [inline base/logging.cc:129]: Here is another copy of `LookupRecordReplaySymbol`?
Code sharing 🙏

File: base/task/task_runner.cc [hunk 4]
```diff
@@ -13,35 +13,42 @@
...
 
+#if !BUILDFLAG(IS_WIN)
 #include <dlfcn.h>
-
-static void (*gRecordReplayRegisterPointerFn)(const char*, void*);
+#else
+#include <windows.h>
+#endif
+
+static void* LookupRecordReplaySymbol(const char* name) {
```
Comment (Domiii) [inline base/task/task_runner.cc:22]: Here is another copy of `LookupRecordReplaySymbol`?

File: third_party/abseil-cpp/absl/base/internal/sysinfo.cc [hunk 5]
```diff
@@ -57,39 +57,37 @@
 #include "absl/base/internal/unscaledcycleclock.h"
 #include "absl/base/thread_annotations.h"
 
+#ifndef _WIN32
 #include <dlfcn.h>
+#endif
 
-static void* gRecordReplayBeginDisallowEventsWithLabelFn;
+static void* LookupRecordReplaySymbol(const char* name) {
```
Comment (Domiii) [inline third_party/abseil-cpp/absl/base/internal/sysinfo.cc:64]: Here is another copy of `LookupRecordReplaySymbol`?

File: third_party/abseil-cpp/absl/base/internal/thread_identity.cc [hunk 6]
```diff
@@ -28,39 +28,39 @@
...
 #include "absl/base/internal/spinlock.h"
 
+#ifndef _WIN32
 #include <dlfcn.h>
+#else
+#include <windows.h>
+#endif
 
-static void* gRecordReplayBeginDisallowEventsWithLabelFn;
+static void* LookupRecordReplaySymbol(const char* name) {
```
Comment (Domiii) [inline third_party/abseil-cpp/absl/base/internal/thread_identity.cc:37]: Here is another copy of `LookupRecordReplaySymbol`?

File: third_party/blink/renderer/platform/heap/thread_state_storage.h [hunk 7]
```diff
@@ -43,12 +43,14 @@ struct ThreadingTrait {
 // Storage for all ThreadState objects. This includes the main-thread
 // ThreadState as well. Keep it outside the class so that PLATFORM_EXPORT
 // doesn't apply to it (otherwise, clang-cl complains).
-//extern thread_local ThreadStateStorage* g_thread_specific_ CONSTINIT
-//    __attribute__((tls_model(BLINK_HEAP_THREAD_LOCAL_MODEL)));
-
+#if BUILDFLAG(IS_WIN)
+extern thread_local ThreadStateStorage* g_thread_specific_ CONSTINIT
+    __attribute__((tls_model(BLINK_HEAP_THREAD_LOCAL_MODEL)));
+#else
```
Comment (Domiii) [inline third_party/blink/renderer/platform/heap/thread_state_storage.h:49]: No `GetThreadStateStorage` in win?

**Review by Domiii — APPROVED** [review]

File: base/functional/callback_helpers.cc (hunk 1 above)
Comment (bhackett1024) [reply base/functional/callback_helpers.cc:17]: It can be pretty difficult when working on this stuff to figure out exactly how to interface with chromium's build system so that we can avoid all code duplication.  I'd rather err on the side of expediency, these are simple stubs and copy-pasting them is an appropriate option to get things to work.  As part of this PR on windows we dynamically look up symbols defined in the same binary in record_replay.cc instead of defining them as externs, because there are many binaries on windows that link against base but not V8.  We could probably do the same thing for linux/mac and eliminate many of these dynamic lookups in other files, but that sort of cleanup isn't appropriate for this issue.

File: third_party/blink/renderer/platform/heap/thread_state_storage.h (hunk 7 above)
Comment (bhackett1024) [reply third_party/blink/renderer/platform/heap/thread_state_storage.h:49]: GetThreadStateStorage is a hack we have on linux to workaround thread_local not working correctly there.  thread_local works correctly on windows, and we should really fix things so that it works on linux too and remove these changes and the other places we have similar workarounds.

File: base/functional/callback_helpers.cc (hunk 1 above)
Comment (Domiii) [reply base/functional/callback_helpers.cc:17]: Makes sense. I added a comment to the "Code Sharing" item in our Runtime 2023 planning doc, for Q3+.
